mlflow/utils/search_utils.py

```diff
@@ -303,7 +229,93 @@ def does_run_match_clause(cls, run, sed):
         else:
             return False
 
-    def filter(self, run):
-        if not self.parsed:
-            self.parsed = self._parse()
-        return all([self.does_run_match_clause(run, s) for s in self.parsed])
+    @classmethod
+    def _parse_search_filter(cls, filter_string):
+        if not filter_string:
+            return []
+        try:
+            parsed = sqlparse.parse(filter_string)
+        except Exception:
+            raise MlflowException("Error on parsing filter '%s'" % filter_string,
+                                  error_code=INVALID_PARAMETER_VALUE)
+        if len(parsed) == 0 or not isinstance(parsed[0], Statement):
+            raise MlflowException("Invalid filter '%s'. Could not be parsed." %
+                                  filter_string, error_code=INVALID_PARAMETER_VALUE)
+        elif len(parsed) > 1:
+            raise MlflowException("Search filter contained multiple expression '%s'. "
+                                  "Provide AND-ed expression list." % filter_string,
+                                  error_code=INVALID_PARAMETER_VALUE)
+        return SearchUtils._process_statement(parsed[0])
+
+    @classmethod
+    def filter(cls, runs, filter_string):
+        """Filters a set of runs based on a search filter string."""
+        if not filter_string:
+            return runs
+        parsed = cls._parse_search_filter(filter_string)
+
+        def run_matches(run):
+            return all([cls._does_run_match_clause(run, s) for s in parsed])
+        return [run for run in runs if run_matches(run)]
+
+    @classmethod
+    def _parse_order_by(cls, order_by):
+        try:
+            parsed = sqlparse.parse(order_by)
+        except Exception:
+            raise MlflowException("Error on parsing order_by clause '%s'" % order_by,
+                                  error_code=INVALID_PARAMETER_VALUE)
+        if len(parsed) != 1 or not isinstance(parsed[0], Statement):
+            raise MlflowException("Invalid order_by clause '%s'. Could not be parsed." %
+                                  order_by, error_code=INVALID_PARAMETER_VALUE)
+
+        statement = parsed[0]
+        if len(statement.tokens) != 1 or not isinstance(statement[0], Identifier):
+            raise MlflowException("Invalid order_by clause '%s'. Could not be parsed." %
+                                  order_by, error_code=INVALID_PARAMETER_VALUE)
+
+        token_value = statement.tokens[0].value
+        is_ascending = True
+        if token_value.lower().endswith(" desc"):
+            is_ascending = False
+            token_value = token_value[0:-len(" desc")]
+        elif token_value.lower().endswith(" asc"):
+            token_value = token_value[0:-len(" false")]
+        identifier = cls._get_identifier(token_value)
+        return (identifier["type"], identifier["key"], is_ascending)
+
+    @classmethod
+    def _get_sort_key(cls, run, key_type, key, ascending):
```

Comment:
This returns a value tuple to embed in sort function. Can we rename function to something like `_get_value_for_sort`

Reply:
I see. It is the "sort key" in the sense that it is the "key" passed to the "sort" function. Maybe I can add a comment indicating the purpose?